### PR TITLE
fix #379

### LINKED ISSUE #379: First entry overwritten with unshift! on a CircularBuffer
I just started getting spurious errors in my OnlineStats builds.  I think this is the culprit.  

```julia
julia> b = CircularBuffer{Float64}(3)
0-element DataStructures.CircularBuffer{Float64}

julia> unshift!(b, 1.0)
1-element DataStructures.CircularBuffer{Float64}:
 1.0

julia> unshift!(b, 2.0)
2-element DataStructures.CircularBuffer{Float64}:
 2.0
 2.15107e-314
```

### PR DESCRIPTION
**Pull request by joshday**
Fix and associated test for #379

### CODE REVIEW

File: src/circular_buffer.jl [hunk 1]
```diff
@@ -72,12 +72,14 @@ function Compat.popfirst!(cb::CircularBuffer)
...
 function Compat.pushfirst!(cb::CircularBuffer, data)
-    # if full, decrement and overwrite, otherwise pushfirst
-    if cb.length < cb.capacity
+    # if full, decrement and overwrite, otherwise unshift
+    if length(cb) == cb.capacity
+        cb.first = (cb.first == 1 ? cb.capacity : cb.first - 1)
+        cb[1] = data
+    else
         cb.length += 1
+        unshift!(cb.buffer, data)
```
Comment (oxinabox) [inline src/circular_buffer.jl]: For 0.7 reasons don't use `unshift!`, use `pushfirst!`  inside the buffer as well

Comment (joshday) [reply src/circular_buffer.jl]: Oops, thanks.

**Comment by oxinabox** [comment]
Would you be able to rebase this?
I am sorry I thought I merged this months ago.

**Comment by joshday** [comment]
No problem, I forgot about this myself.  

I got myself in git hell though with this PR...new PR coming in a sec.

**Pull request closed without merging**